crates/bls-crypto/src/bls/public.rs

```diff
@@ -216,25 +169,131 @@ impl CanonicalDeserialize for PublicKey {
+            let x = Fq2::read(x_bytes_with_y.as_slice())?;
+            let x3b = <Bls12_377G2Parameters as SWModelParameters>::add_b(
+                &((x.square() * &x) + &<Bls12_377G2Parameters as SWModelParameters>::mul_by_a(&x)),
+            );
+            let y = x3b.sqrt().ok_or(io::Error::new(
+                io::ErrorKind::NotFound,
+                "couldn't find square root for x",
+            ))?;
+
+            let y_c0_big = y.c0.into_repr();
+            let y_c1_big = y.c1.into_repr();
+
+            let negy = -y;
+
+            let (bigger, smaller) = {
+                let half = Fq::modulus_minus_one_div_two();
+                if y_c1_big > half {
+                    (y, negy)
+                } else if y_c1_big == half && y_c0_big > half {
+                    (y, negy)
+                } else {
+                    (negy, y)
+                }
+            };
+
+            let chosen_y = if y_over_half { bigger } else { smaller };
+            Ok((x, chosen_y))
+        };
+
         let rng = &mut thread_rng();
-        for _ in 0..100 {
+        let mut i = 0;
+        loop {
+            let sk = PrivateKey::generate(rng);
+            let pk = sk.to_public();
+            if pk.as_ref().into_affine().y.c1 == Fq::zero() {
+                continue;
+            }
+
+            let mut pk_bytes = vec![];
+            pk.write(&mut pk_bytes).unwrap();
+
+            let mut pk_bytes2 = vec![];
+            pk.serialize(&mut pk_bytes2).unwrap();
+
+            let mut pk_bytes3 = vec![];
+            old_serialization_logic(pk, &mut pk_bytes3).unwrap();
+
+            assert_eq!(pk_bytes, pk_bytes2);
+            assert_eq!(pk_bytes, pk_bytes3);
+
+            let de_pk = PublicKey::read(&pk_bytes[..]).unwrap();
+            let de_pk2 = PublicKey::deserialize(&mut &pk_bytes[..]).unwrap();
+
+            assert_eq!(de_pk, de_pk2);
+
+            // check that the points match (the PartialEq does only bytes equality)
+            assert_eq!(de_pk.as_ref().x, de_pk2.as_ref().x);
+            assert_eq!(de_pk.as_ref().y, de_pk2.as_ref().y);
+
+            i += 1;
+            if i == 1000 {
+                break;
+            }
+        }
+
+        // check cases where c1 = 0
+        for _ in 0..1000 {
             let sk = PrivateKey::generate(rng);
             let pk = sk.to_public();
+            let mut pk_affine = pk.as_ref().into_affine();
+            pk_affine.y.c1 = Fq::zero();
+            let pk = PublicKey::from(pk_affine.into_projective());
 
             let mut pk_bytes = vec![];
             pk.write(&mut pk_bytes).unwrap();
 
             let mut pk_bytes2 = vec![];
             pk.serialize(&mut pk_bytes2).unwrap();
 
+            let mut pk_bytes3 = vec![];
+            old_serialization_logic(pk, &mut pk_bytes3).unwrap();
+
             assert_eq!(pk_bytes, pk_bytes2);
+            // check for the bug mentioned in https://github.com/celo-org/bls-zexe/issues/149
```

Comment:
IMO add the full description of the issue here, it's not really clear what the bug is if you skim #149